ruma-appservice-api/src/event/push_events/v1.rs

```diff
@@ -41,6 +41,54 @@ impl IncomingRequest {
     pub fn new(txn_id: String, events: Vec<Raw<AnyEvent>>) -> IncomingRequest {
         IncomingRequest { txn_id, events }
     }
+
+    /// Consumes the `IncomingRequest` and tries to convert it to a `sync_events::Response`
+    ///
+    /// This is a helper conversion in cases where it's easier to work with `sync_events::Response`
+    /// instead of the original `push_events::IncomingRequest`. It puts all events with a `room_id`
+    /// into the `JoinedRoom`'s `timeline`. The rationale behind that is that incoming Appservice
+    /// transactions from the homeserver are not necessarily bound to a specific user but can cover
+    /// a multitude of namespaces, and as such the Appservice basically only "observes joined
+    /// rooms".
+    #[cfg(feature = "helper")]
+    pub fn try_into_sync_response(
+        self,
+        next_batch: impl Into<String>,
+    ) -> Result<ruma_client_api::r0::sync::sync_events::Response, serde_json::Error> {
+        use ruma_client_api::r0::sync::sync_events;
+        use ruma_identifiers::RoomId;
+        use serde::{de::IgnoredAny, Deserialize};
+
+        #[derive(Debug, Deserialize)]
+        struct EventDeHelper {
+            #[serde(rename = "type")]
+            ev_type: String,
+            event_id: Option<IgnoredAny>,
+            room_id: Option<RoomId>,
+        }
+
+        let mut response = sync_events::Response::new(next_batch.into());
+
+        for raw_event in self.events {
+            let helper: EventDeHelper = serde_json::from_str(raw_event.json().get())?;
+            let event_json = Raw::into_json(raw_event);
+
+            if let Some(room_id) = helper.room_id {
+                let join = response.rooms.join.entry(room_id).or_default();
+
+                // Currently homeservers only push PDUs to appservices, no EDUs
+                //
+                // There's the open MSC2409 regarding supporting EDUs in the future, though it
+                // seems to be planned to put EDUs into a different JSON key:
+                // https://github.com/matrix-org/matrix-doc/pull/2409
+                if helper.event_id.is_some() {
```

Comment:
Do we need this check? If we expect there to be PDUs only, it seems better to construct a faulty sync response from a faulty `push_events::IncomingRequest` and have the consumer handle that, no?

Reply:
Rather than just throwing away the event entirely. I'd also like to see a `tracing::warn!`ing logged when an event doesn't contain a room ID.

Reply:
> it seems better to construct a faulty sync response from a faulty push_events::IncomingRequest and have the consumer handle that, no?

Sounds fair, changed

> I'd also like to see a tracing::warn!ing logged when an event doesn't contain a room ID.

Done

Also moved the comment into the method docs